samples/MemoryStateStoreSample/Program.cs

```diff
@@ -0,0 +1,17 @@
+using MemoryStateStoreSample.Services;

```

Comment:
I'm missing the ability to "Register" many components as the user want. It's not a common case but it's possible

[See this](https://github.com/dapr-sandbox/components-go-sdk/blob/main/examples/mesh/main.go#L32)

It's an example of a single process creating two sockets files one for each component.

Do you think we should have something similar here? A sort of a `ComponentRegistry` ?

Reply:
My current model allows multiple components to be registered (on a single socket) that do not overlap in function.  I'm not opposed to multiplexing in your sense, though I'm not convinced how prevalent it would be in reality.  For example, how likely is a customer to want to create just a single pluggable component application that contains multiple components that overlap in function (such that multiple sockets are required) versus creating multiple pluggable component applications?  Even if the former is possible, it feels like the latter would be the better way to do it.

Another question is whether we expect developers to create a single implementation that expects to serve as multiple types of component (e.g. implements the interfaces for state store *and* binding *and* pub-sub) or whether we expect developers to create separate implementations for each component interface.  The latter feels like the better approach from a separation of concerns point of view, but do we want to support the former as well?  (My model currently doesn't; how opinionated do we want our SDKs to be?)

Reply:
Think about a single vendor maintaining a bunch of components.

Let’s say that in the future, pluggable components are a perfect replacement for built-in ones and we decide to migrate them.

Now think how many azure components we do have, same for aws, they have different building blocks being implemented under different names.

Im not saying that’s the way to go, but they could ship them all as a single container instead of managing 10+ different containers/pluggable components.

so we can have the “aws” pluggable component mesh, notice that there could be a lot of resources being shared like auth client instances and database connections that does not need to be duplicated because they live in the same process.

Obviously I’m simplifying things a bit here, but I really like the idea of a single image for many containers to allow things like: ship the entire components contrib as a single container, I did this myself a while ago and it worked perfectly.

Reply:
I'll buy a container being a ship vehicle for a bundle of related components, and being able to decide, at run-time, which of those to instantiate/register as gRPC services.  I'm not as certain about the practicality of always registering and/or instantiating a large number of components within the same container as opposed to spreading them across several different container instances; the latter would seem better from a resiliency standpoint. 

That said, I'll look into how requests from specific socket endpoints could be routed to specific services.

Reply:
Single container with multiple components does not block scenario for multiple containers to be honest. I think this is a superset scenario

Reply:
I love that piece of code you pointed on [mesh/main.go](https://github.com/dapr-sandbox/components-go-sdk/blob/main/examples/mesh/main.go#L32).

But on practical terms, is this use case (multiple plugins, each exposing distinct building blocks, hosted by a single pluggable component _container_) one we want the SDKs to support? It looks hella-cool (to sound like a 90's teen) but a few threads before we were talking about how opinionated the SDK ought to be.

Yes, we want the SDKs to have feature parity, but we need to be consistent on what we want to enable (or not) with the SDKs. Otherwise we need to revisit the possibility of exposing internals of the container back to the user and whatnot.

Reply:
I spent some time looking at multi-socket support and pushed something that provides it in a similar manner to Go.  While I don't expect users to expose dozens of binding components from a dozen sockets in the same instance of a container, it does offer a reasonable amount of flexibility and is useful even if only to be consistent across SDKs.

This allows something like this, which I think fits the convention and style of ASP.NET reasonably well:

```csharp
var app = DaprPluggableComponentsApplication.Create();

app.RegisterService(
    "<socket A>",
    serviceBuilder =>
    {
        // Register a state store with the service.
        serviceBuilder.RegisterStateStore<StateStoreA>();
    });

app.RegisterService(
    "<socket B>",
    serviceBuilder =>
    {
        // Register a state store with the service.
        serviceBuilder.RegisterStateStore<StateStoreB>();
    });

app.Run();

class StateStoreA : IStateStore
{
    // ...
}

class StateStoreB : IStateStore
{
    // ...
}
```

Reply:
lgtm